File: src/main/java/no/nav/syfo/kafka/mapper/SoknadDtoToSykepengesoknadMapper.java
package no.nav.syfo.kafka.mapper;

import no.nav.syfo.domain.dto.*;
import no.nav.syfo.kafka.soknad.dto.SoknadDTO;
import no.nav.syfo.kafka.soknad.dto.SoknadPeriodeDTO;
import no.nav.syfo.kafka.soknad.dto.SporsmalDTO;
import no.nav.syfo.kafka.soknad.dto.SvarDTO;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Optional.ofNullable;

public final class SoknadDtoToSykepengesoknadMapper {

    private static <T extends Enum<T>> T konverter(Class<T> tClass, String name) {
        return ofNullable(name).map(n -> Enum.valueOf(tClass, n)).orElse(null);
    }

    private static Svar konverter(SvarDTO svar) {
        return Svar.builder()
                .verdi(svar.getVerdi())
                .build();
    }

    private static Sporsmal konverter(SporsmalDTO sporsmal) {
        return Sporsmal.builder()
                .id(sporsmal.getId())
                .tag(sporsmal.getTag())
                .sporsmalstekst(sporsmal.getSporsmalstekst())
                .undertekst(sporsmal.getUndertekst())
                .svartype(konverter(Svartype.class, sporsmal.getSvartype()))
                .min(sporsmal.getMin())
                .max(sporsmal.getMax())
                .kriterieForVisningAvUndersporsmal(konverter(Visningskriterie.class, sporsmal.getKriterieForVisningAvUndersporsmal()))
                .svar(stream(sporsmal.getSvar())
                        .map(SoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .undersporsmal(stream(sporsmal.getUndersporsmal())
                        .map(SoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .build();
    }

    private static SoknadPeriode konverter(SoknadPeriodeDTO soknadPeriode) {
        return SoknadPeriode.builder()
                .fom(soknadPeriode.getFom())
                .tom(soknadPeriode.getTom())
                .grad(soknadPeriode.getGrad())
                .build();
    }

    public static Sykepengesoknad konverter(SoknadDTO soknad) {
        return Sykepengesoknad.builder()
                .id(soknad.getId())
                .sykmeldingId(soknad.getSykmeldingId())
                .aktorId(soknad.getAktorId())
                .soknadstype(konverter(Soknadstype.class, soknad.getSoknadstype()))
                .status(soknad.getStatus())
                .fom(soknad.getFom())
                .tom(soknad.getTom())
                .opprettet(soknad.getOpprettetDato().atStartOfDay())
                .sendtNav(ofNullable(soknad.getInnsendtDato()).map(LocalDate::atStartOfDay).orElse(null))
                .arbeidsgiver(soknad.getArbeidsgiver())
                .arbeidssituasjon(konverter(Arbeidssituasjon.class, soknad.getArbeidssituasjon()))
                .startSykeforlop(soknad.getStartSykeforlop())
                .sykmeldingSkrevet(ofNullable(soknad.getSykmeldingUtskrevet()).map(LocalDate::atStartOfDay).orElse(null))
                .korrigertAv(soknad.getKorrigertAv())
                .korrigerer(soknad.getKorrigerer())
                .soknadPerioder(stream(soknad.getSoknadPerioder())
                        .map(SoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .sporsmal(stream(soknad.getSporsmal())
                        .map(SoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .build();
    }

    private static <T> Stream<T> stream(List<T> list) {
        return list == null ? Stream.empty() : list.stream();
    }
}

File: src/main/kotlin/no/nav/syfo/kafka/mapper/SoknadDtoToSykepengesoknadMapper.kt
package no.nav.syfo.kafka.mapper

import no.nav.syfo.domain.dto.SoknadPeriode
import no.nav.syfo.domain.dto.Sporsmal
import no.nav.syfo.domain.dto.Svar
import no.nav.syfo.domain.dto.Sykepengesoknad
import no.nav.syfo.kafka.felles.SoknadsperiodeDTO
import no.nav.syfo.kafka.felles.SporsmalDTO
import no.nav.syfo.kafka.felles.SvarDTO
import no.nav.syfo.kafka.soknad.dto.SoknadDTO


private inline fun <reified U : Enum<*>> String?.stringEnumValueOrNull(): U? =
        U::class.java.enumConstants.firstOrNull { it.name == this }


private inline fun <T : Enum<*>, reified U : Enum<*>> T?.enumValueOrNull(): U? =
        U::class.java.enumConstants.firstOrNull { it.name == this?.name }

private fun SporsmalDTO.toSporsmal(): Sporsmal =
        Sporsmal(
                id = id!!,
                tag = tag!!,
                sporsmalstekst = sporsmalstekst,
                undertekst = undertekst,
                svartype = svartype.enumValueOrNull(),
                min = min,
                max = max,
                kriterieForVisningAvUndersporsmal = kriterieForVisningAvUndersporsmal.enumValueOrNull(),
                svar = svar?.map { it.toSvar() },
                undersporsmal = undersporsmal?.map { it.toSporsmal() }
        )

private fun SvarDTO.toSvar(): Svar =
        Svar(verdi)

private fun SoknadsperiodeDTO.toSoknadPeriode(): SoknadPeriode =
        SoknadPeriode(
                fom = fom,
                tom = tom,
                grad = grad,
                faktiskGrad = faktiskGrad)

fun SoknadDTO.toSykepengesoknad(): Sykepengesoknad =
        Sykepengesoknad(
                id = id!!,
                sykmeldingId = sykmeldingId,
                aktorId = aktorId!!,
                soknadstype = soknadstype.stringEnumValueOrNull(),
                status = status!!,
                fom = fom,
                tom = tom,
                opprettet = opprettetDato!!.atStartOfDay(),
                sendtNav = innsendtDato?.atStartOfDay(),
                arbeidsgiver = arbeidsgiver,
                arbeidssituasjon = arbeidssituasjon.stringEnumValueOrNull(),
                startSykeforlop = startSykeforlop,
                sykmeldingSkrevet = sykmeldingUtskrevet?.atStartOfDay(),
                korrigertAv = korrigertAv,
                korrigerer = korrigerer,
                soknadPerioder = soknadPerioder?.map { it.toSoknadPeriode() },
                sporsmal = sporsmal!!.map { it.toSporsmal() },
                avsendertype = avsendertype?.name?.stringEnumValueOrNull(),
                egenmeldtSykmelding = egenmeldtSykmelding,
                harRedusertVenteperiode = harRedusertVenteperiode ?: false
        )


